### PR TITLE
Refactor devices, add cover size stuff

### PR DESCRIPTION
**Pull request by pgaskin**
* A lot more information (codenames, storage, etc)
* Cover sizes
* Cleaner code, easier to extend in the future
* Codenames now match the internal layout (class -> family + optional secondary)
* Based on firmware 4.18.13737

Also see https://gist.github.com/geek1011/613b34c23f026f7c39c50ee32f5e167e and https://github.com/shermp/Kobo-UNCaGED/issues/16

cc @shermp, @NiLuJe, @davidfor

### CODE REVIEW

**Review by pgaskin — COMMENTED** [review]
TODO:
- [x] Write more tests
- [x] Add more godoc comments
- [x] Update my tools which depend on this
- [x] Test cover generation (work will be in the covergen branch of kepubify)

Also, this is mostly ready for review now.

**Comment by shermp** [comment]
Have you thought about having a generic "fallback" device, name etc with safe defaults such that new devices can still use the library, even if they aren't supported yet?

Also, this looks awesome. I'm inclined to switch to this library in Kobo-UNCaGED to avoid duplication of code.

**Comment by pgaskin** [comment]
OK, I've tested the cover generation and some of the device detection as part of geek1011/kepubify#44, and everything seems to work well.

File: kobo/device.go [hunk 1]
```diff
@@ -74,7 +74,7 @@ const (
 
 // Cover types.
 const (
-	CoverTypeLibFull CoverType = "N3_FULL"
+	CoverTypeLibFull CoverType = "N3_LIBRARY_FULL"
```
Comment (NiLuJe) [inline kobo/device.go:95]: IIRC, actually, it's both.

`N3_FULL` is the actual screen-sized one (sleep screen), while `N3_LIBRARY_FULL` is the one used in the Details view.

**Comment by shermp** [comment]
> It's Qt4's QtHash, FWIW ;).
> 
> (c.f., https://github.com/NiLuJe/kfmon/blob/f19f8582e7b53d14b7a60cc5b8bc43bac1930b48/kfmon.c#L1096)



> Ok, I was wondering how much time @shermp would have spent reversing this... 😃
> 
> I'll add a note about that for future reference.

Yeah, I lifted it straight from the Kobo driver in Calibre, which is where I originally saw it.

EDIT: Somewhere in our refactors, that little nugget of info was lost. My original comment was:
```
// genImagePath generates the directory structure used by
// kobo to store the cover image files.
// It has been ported from the implementation in the KoboTouch
// driver in Calibre
func genImageDirPath(imageID string) string {
```

File: kobo/device.go (hunk 1 above)
Comment (pgaskin) [reply kobo/device.go:95]: Are the sizes the same, or do I need to add another set of sizes?

Comment (pgaskin) [reply kobo/device.go:95]: Nevermind, I just realized I was diverging from my notes when I changed those around. I'll fix the sizes.

File: kobo/device.go [hunk 2]
```diff
@@ -325,9 +328,9 @@ func (c CoverType) String() string {
 // Resize returns the dimensions to resize sz to for the cover type and target size.
 func (c CoverType) Resize(target image.Point, sz image.Point) image.Point {
 	switch c {
-	case CoverTypeLibList:
+	case CoverTypeFull:
 		return resizeKeepAspectRatio(sz, target, false)
-	case CoverTypeLibFull, CoverTypeLibGrid:
+	case CoverTypeLibFull, CoverTypeLibGrid, CoverTypeLibList:
 		return resizeKeepAspectRatio(sz, target, true)
 	}
```
Comment (pgaskin) [inline kobo/device.go]: @NiLuJe, how does this look now?

P.S. That liblist one at the top was a mistake I made earlier while refactoring.

Comment (NiLuJe) [reply kobo/device.go]: Yup, looks good now ;).

**Comment by pgaskin** [comment]
Unless anyone has more suggestions/objections, I'm planning to merge this after I write tests for the cover stuff.

**Comment by pgaskin** [comment]
I've released this as v2.

**Comment by pgaskin** [comment]
@shermp, are you planning on updating KU yourself?

**Comment by shermp** [comment]
I was thinking of pulling in koboutils as a library and using it instead.

No need to duplicate code/effort :)

**Comment by pgaskin** [comment]
That was what I meant, but were you planning to implement it?

**Comment by shermp** [comment]
Yeah. Want to finalize the changes to UNCaGED first before updating KU.

**Comment by pgaskin** [comment]
Sounds good.

**Comment by shermp** [comment]
Speaking of UNCaGED, would you be willing to have a look at the open PR at some point? I left it for a while, seeing as you were taking a break from Kobo stuff.

**Comment by pgaskin** [comment]
Yep, will do today or tomorrow. I'll be taking another break afterwards until the next firmware release.

**Comment by shermp** [comment]
Thanks

**Pull request merged**